File: app/src/main/java/home/oleg/popularmovies/presentation/list/presenter/ListPresenter.java
package home.oleg.popularmovies.presentation.list.presenter;

/**
 * Created by Oleg on 15.04.2017.
 */

public interface ListPresenter {
    void fetchMovies(Filter filter);

    enum Filter {
        TOP_RATED("top_rated"),
        POPULAR("popular"),
        FAVOURITE("favourite");

        private String value;

        Filter(String value) {
            this.value = value;
        }

        public static Filter resolve(String filter){
            switch (filter){
                case "top_rated":
                    return TOP_RATED;
                case "popular":
                    return POPULAR;
                case "favourite":
                    return FAVOURITE;
                default:
                    return POPULAR;
            }
        }

        public String getValue() {
           return this.value;
        }
    }
}

File: app/src/main/java/home/oleg/popularmovies/presentation/list/presenter/ListPresenter.kt
package home.oleg.popularmovies.presentation.list.presenter

import home.oleg.popularmovies.domain.MovieRepository
import home.oleg.popularmovies.presentation.list.ListView
import home.oleg.popularmovies.presentation.mappers.MovieToMovieViewModelMapper
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.*
import javax.inject.Inject

/**
 * Created by Oleg on 15.04.2017.
 */

class ListPresenter @Inject constructor(
        private val disposableBag: CompositeDisposable,
        private val mapper: MovieToMovieViewModelMapper,
        private val movieRepository: MovieRepository) {

    private var view: ListView? = null

    fun attachView(view: ListView) {
        this.view = view

        /*Observable.interval(5000, TimeUnit.MILLISECONDS).flatMap {
            movieRepository.getMovies(MovieRepository.Filter.TOP_RATED)

        }.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread(), true)
                .subscribe({
                }, {
                    it.printStackTrace()
                })
                .let(disposableBag::add)*/
    }

    fun fetchMovies(filter: MovieRepository.Filter) {
        movieRepository.getMovies(filter)
                .map { it.map(mapper) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread(), true)
                .doOnSubscribe { view?.showLoading() }
                .doFinally { view?.hideLoading() }
                .subscribe({
                    view?.fillList(it)
                }, {
                    it.printStackTrace()
                    view?.showError()
                })
                .let(disposableBag::add)
    }

    fun detachView() {
        view = null
        disposableBag.clear()
    }
}